tsl/src/fdw/relinfo.c

```diff
@@ -308,74 +245,97 @@ estimate_tuples_and_pages_using_shared_buffers(PlannerInfo *root, Hypertable *ht
-
-	/* Let's first try figuring out number of tuples/pages using stats from previous chunks,
-	otherwise make an estimation based on shared buffers size */
-	estimates = estimate_tuples_and_pages_using_prev_chunks(root, hyperspace, chunk);
-	if (estimates->tuples <= 0 || estimates->pages == 0)
-		estimates = estimate_tuples_and_pages_using_shared_buffers(root, ht, rel->reltarget->width);
-
-	chunk_fillfactor = estimate_chunk_fillfactor(chunk, hyperspace);
-	/* adjust tuples/pages using chunk_fillfactor */
-	rel_estimates_apply_fillfactor(estimates, chunk_fillfactor);
-	set_rel_estimates(rel, estimates);
+	RelOptInfo *parent_info = root->simple_rel_array[parent_relid];
+	TsFdwRelInfo *parent_private = fdw_relinfo_get(parent_info);
+	RangeTblEntry *parent_rte = planner_rt_fetch(parent_relid, root);
+	Cache *hcache = ts_hypertable_cache_pin();
+	Hypertable *ht = ts_hypertable_cache_get_entry(hcache, parent_rte->relid, CACHE_FLAG_NONE);
+	Hyperspace *hyperspace = ht->space;
+	RangeTblEntry *chunk_rte = planner_rt_fetch(chunk_rel->relid, root);
+	Chunk *chunk = ts_chunk_get_by_relid(chunk_rte->relid, true /* fail_if_not_found */);
+
+	const double fillfactor = estimate_chunk_fillfactor(chunk, hyperspace);
+	const bool have_chunk_statistics = chunk_rel->pages != 0 || chunk_rel->tuples > 0;
+	const bool have_moving_average =
+		parent_private->average_chunk_pages != 0 || parent_private->average_chunk_tuples > 0;
+	if (!have_chunk_statistics)
+	{
+		/*
+		 * If we don't have the statistics from ANALYZE for this chunk,
+		 * use the moving average of chunk sizes. If we don't have even
+		 * that, use an estimate based on the default shared buffers
+		 * size for a chunk.
+		 */
+		if (have_moving_average)
+		{
+			chunk_rel->pages = parent_private->average_chunk_pages * fillfactor;
+			chunk_rel->tuples = parent_private->average_chunk_tuples * fillfactor;
+		}
+		else
+		{
+			estimate_tuples_and_pages_using_shared_buffers(root, ht, chunk_rel);
+			chunk_rel->pages *= fillfactor;
+			chunk_rel->tuples *= fillfactor;
+		}
+	}
+
+	if (!have_moving_average)
+	{
+		/*
+		 * Initialize the moving average data if we don't have any yet.
+		 * Use even a bad estimate from shared buffers, to save on
+		 * recalculating the same bad estimate for the subsequent chunks
+		 * that are likely to not have the statistics as well.
+		 */
+		parent_private->average_chunk_pages = chunk_rel->pages;
+		parent_private->average_chunk_tuples = chunk_rel->tuples;
+	}
+	else if (have_chunk_statistics)
+	{
+		/*
+		 * We have the moving average of chunk sizes and a good estimate
+		 * of this chunk size from ANALYZE. Update the moving average.
+		 */
+		const double f = 0.1;
+		parent_private->average_chunk_pages =
+			(1 - f) * parent_private->average_chunk_pages + f * chunk_rel->pages / fillfactor;
+		parent_private->average_chunk_tuples =
+			(1 - f) * parent_private->average_chunk_tuples + f * chunk_rel->tuples / fillfactor;
```

Comment:
It seems as if we update the moving average on every query that involves the chunk instead of when we have new stats. Won't it make the moving average a bit weird if we are, e.g., querying the same chunk many times?

In other words, a chunk's effect on the moving average depends on how many times it is queried rather than just once.

Ideally, one would expect to add a chunk to the moving average only once, or at least only when it is re-analyzed.

Does it make more sense to update this moving average when we run ANALYZE instead of at query time?

We also have to account for the fact that a chunk is sometimes "incomplete", but i guess that is what the fillfactor is supposed to account for.